File: 백준/Silver/2644. 촌수계산/촌수계산.java
import java.util.Scanner;

public class Main {

    private static int[][] graph;

    private static boolean[] visited;

    private static int b;

    private static int result;

    public static void main(String[] args) {
        try (Scanner scanner = new Scanner(System.in)) {
            int n = scanner.nextInt();
            int a = scanner.nextInt();
            b = scanner.nextInt();
            int m = scanner.nextInt();

            graph = new int[n][n];
            visited = new boolean[n];

            for (int i = 0; i < m; i++) {
                int x = scanner.nextInt() - 1;
                int y = scanner.nextInt() - 1;

                graph[x][y] = 1;
                graph[y][x] = 1;
            }

            dfs(a - 1);

            System.out.println(result == 0 ? -1 : result);
        }
    }

    private static int dfs(int n) {
        int count = 1;

        if (n == b - 1) {

            int cc = 0;
            for (boolean b : visited) {
                if (b) {
                    cc++;
                }
            }
            result = cc;
        }

        for (int i = 0; i < graph.length; i++) {
            if (!visited[i] && graph[n][i] == 1) {
                visited[n] = true;
                count += dfs(i);
                visited[n] = false;
            }
        }
        return count;
    }
}

File: 백준/Silver/2644. 촌수계산/촌수계산.kt
var result: Int = -1

fun main() {
    val n = readln().toInt()
    val graph = Array(n) { BooleanArray(n) }
    val visited = BooleanArray(n)

    val (a, b) = readln().split(" ").map { it.toInt() }

    val m = readln().toInt()
    repeat(m) {
        val (x, y) = readln().split(" ").map { it.toInt() }
        graph[x - 1][y - 1] = true
        graph[y - 1][x - 1] = true
    }

    val sum = 1
    for (i in graph.indices) {
        if (graph[a - 1][i]) {
            dfs(i, b, graph, visited, sum)
        }
    }

    println(result)
}

private fun dfs(
    num: Int,
    b: Int,
    graph: Array<BooleanArray>,
    visited: BooleanArray,
    sum: Int,
): Boolean {
    visited[num] = true
    if (b == num + 1) {
        result = sum
    }


    for (i in graph.indices) {
        if (graph[num][i] && !visited[i]) {
            dfs(i, b, graph, visited, sum + 1)
        }
    }

    return false
}